server/src/main/java/io/druid/client/HttpServerInventoryView.java

```diff
@@ -256,8 +253,9 @@ public void registerSegmentCallback(
       Executor exec, SegmentCallback callback, Predicate<Pair<DruidServerMetadata, DataSegment>> filter
   )
   {
-    segmentCallbacks.put(callback, exec);
-    segmentPredicates.put(callback, filter);
+    SegmentCallback filteringSegmentCallback = new SingleServerInventoryView.FilteringSegmentCallback(callback, filter);
+    segmentCallbacks.put(filteringSegmentCallback, exec);
+    segmentPredicates.put(filteringSegmentCallback, filter);
```

Comment:
Looks that ```segmentPredicates``` is not needed if ```FilteringSegmentCallback``` has a method like ```getFilter()```.

Reply:
it is more used in updating `finalPredicate` in [register/run]SegmentCallback, it was mostly copied from `BatchServerInventoryView` . i guess, we could do something smarter to remove this field but those predicates do need to be stored in filteringSegmenCallback with a new type as SegmentCallback can't have those.
i have tested this code to a decent degree and would keep it like this for now unless there was a major advantage, also intending to just fix some bugs in this pr.

Reply:
Ok.